File: firebase/functions/lib/admin/payments/analytics_util.dart
import 'dart:async';
import 'dart:convert';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'package:data_models/analytics/analytics_entities.dart';
import 'package:node_http/node_http.dart' as http;

AnalyticsUtil analyticsUtil = AnalyticsUtil();

class AnalyticsUtil {
  String get _secretKey => functions.config.get('segment.write_key') as String;

  /// Call to track a user event
  void logEvent({required String userId, required AnalyticsEvent event}) {
    _doLogEvent(userId: userId, event: event);
  }

  Future<void> _doLogEvent({
    required String userId,
    required AnalyticsEvent event,
  }) async {
    final encodedKey = base64.encode(('$_secretKey:').codeUnits);
    final headers = {
      'Authorization': 'Basic $encodedKey',
      "Content-Type": 'application/json',
    };

    final body = {
      'event': event.getEventType(),
      'properties': event.toJson(),
      'userId': userId,
    };

    final response = await http.post(
      Uri.parse('https://api.segment.io/v1/track'),
      headers: headers,
      body: jsonEncode(body),
    );

    if (response.statusCode > 299) {
      print('Segment POST error:');
      print(response.body);
    }
  }
}

File: firebase/functions/src/admin/payments/analytics_util.ts
import * as functions from 'firebase-functions';
import * as https from 'https';

export const analyticsUtil = {
  get _secretKey(): string {
    return functions.config().segment?.write_key as string;
  },

  logEvent({ userId, event }: { userId: string; event: Record<string, unknown> }): void {
    this._doLogEvent({ userId, event }).catch(console.error);
  },

  async _doLogEvent({ userId, event }: { userId: string; event: Record<string, unknown> }): Promise<void> {
    const encodedKey = Buffer.from(`${analyticsUtil._secretKey}:`).toString('base64');
    const body = JSON.stringify({
      event: (event as any).eventType ?? (event as any).type ?? 'unknown',
      properties: event,
      userId,
    });

    return new Promise((resolve, reject) => {
      const req = https.request(
        {
          hostname: 'api.segment.io',
          path: '/v1/track',
          method: 'POST',
          headers: {
            'Authorization': `Basic ${encodedKey}`,
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(body),
          },
        },
        (res) => {
          let data = '';
          res.on('data', (chunk) => (data += chunk));
          res.on('end', () => {
            const statusCode = res.statusCode ?? 500;
            if (statusCode > 299) {
              console.error('Segment POST error:', data);
            }
            resolve();
          });
        }
      );
      req.on('error', (e) => { console.error(e); resolve(); });
      req.write(body);
      req.end();
    });
  },
};
